Rework goPackages

kamilchm opened the pull request:
###### Motivation for this change

This is my try on better `goPackages` as I mentioned in https://github.com/NixOS/nixpkgs/pull/13819#issuecomment-195839153 

I started with an assumption that there's no need to have `goPackages` that aren't used in a compiled program. So I've created new derivations in proper program categories for all `all-packages.nix` that comes from `go-packages.nix` using prototype tool `go2nix`.
All program dependencies are extracted to json file than can include other json dependency files for better package reuse and making overrides possible.
`libs.json` and `deps.json` files can be easily updated with tools like proposed in https://github.com/NixOS/nixpkgs/pull/13819 or generated from go native dependency managers.

There are few more things to do, but I'm looking for feedback to proritize tasks. Let's make go packaging better.

CC: @edolstra @cstrahan @copumpkin @viric @hrdinka @wkennington @zimbatm 
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

kamilchm committed:
First programs extracted from go-packages: deis and gawp

kamilchm committed:
When using common deps include only specified packages, not all set

kamilchm committed:
Reusing go libs in deis build

kamilchm committed:
asciinema: top level pkgs extracted from goPackages

kamilchm committed:
fzf: application extracted from goPackages

kamilchm committed:
interlock: extracted from goPackages

kamilchm committed:
mongodb-tools: extracted from goPackages

kamilchm committed:
flannel: extracted from goPackages

kamilchm committed:
git-lfs: extracted from goPackages

kamilchm committed:
go-mtpfs: extracted from goPackages

kamilchm committed:
go-pup: extracted from goPackages

kamilchm committed:
go-sct: extracted from goPackages

kamilchm committed:
go-upower-notify: extracted from goPackages

kamilchm committed:
lxd: extracted from goPackages

kamilchm committed:
ngrok: extracted from goPackages

kamilchm committed:
s3gof3r: extracted from goPackages

kamilchm committed:
skydns: extracted from goPackages

kamilchm committed:
wego: extracted from goPackages

kamilchm committed:
textql: extracted from goPackages

kamilchm committed:
oh: extracted from goPackages

kamilchm committed:
go-repo-root: extracted from goPackages

kamilchm committed:
gox: extracted from goPackages

kamilchm committed:
mesos-dns: extracted from goPackages

kamilchm committed:
gotty: extracted from goPackages

kamilchm committed:
vimPlugins: FIX fzf outside goPackages

kamilchm committed:
etcd: extracted from goPackages

kamilchm committed:
nsq: extracted from goPackages

kamilchm committed:
oauth2_proxy: extracted from goPackages

kamilchm committed:
serfdom: extracted from goPackages

kamilchm committed:
gocode: extracted from goPackages

kamilchm committed:
gotags: extracted from goPackages

kamilchm committed:
golant: extracted from goPackages

kamilchm committed:
goimports: extracted from goPackages

kamilchm committed:
i3cat: extracted from goPackages

kamilchm committed:
hologram: extracted from goPackages

kamilchm committed:
golint: FIX proper link to godeps file

kamilchm committed:
confd: extracted from goPackages

kamilchm committed:
consul: extracted from goPackages

kamilchm committed:
consul-alerts: extracted from goPackages

kamilchm committed:
consul--template: extracted from goPackages

kamilchm committed:
vault: extracted from goPackages

kamilchm committed:
consul: FIX consul-ui reference

kamilchm committed:
syncthing012: extracted from goPackages

kamilchm committed:
xmpp-client: extracted from goPackages

kamilchm committed:
caddy: extracted from goPackages

kamilchm committed:
glide: extracted from goPackages

kamilchm committed:
terraform: extracted from goPackages

kamilchm committed:
pond: extracted from goPackages

kamilchm committed:
prometheus: extracted from goPackages

kamilchm committed:
prom2json: extracted from goPackages

kamilchm committed:
prometheus-alertmanager: extracted from goPackages

kamilchm committed:
prometheus-cli: extracted from goPackages

kamilchm committed:
goDeps: support for mercurial repos

kamilchm committed:
prometheus-collectd-exporter: extracted from goPackages

kamilchm committed:
prometheus-haproxy-exporter: extracted from goPackages

kamilchm committed:
prometheus-mesos-exporter: extracted from goPackages

kamilchm committed:
prometheus-mysqld-exporter: extracted from goPackages

kamilchm committed:
prometheus-nginx-exporter: extracted from goPackages

kamilchm committed:
prometheus-node-exporter: extracted from goPackages

kamilchm committed:
prometheus-pushgateway: extracted from goPackages

kamilchm committed:
prometheus-statsd-bridge: extracted from goPackages

kamilchm committed:
go2nix: extracted from goPackages

kamilchm committed:
govers: extracted from goPackages

kamilchm committed:
ngrok: FIX go-bindata moved to all-packages

kamilchm committed:
drive: extracted from goPackges

kamilchm committed:
no more goPackages

kamilchm committed:
ipfs: extracted from goPackages

kamilchm committed:
FIX: deis and gawp shas

kamilchm committed:
prometheus-alertmanager: fetchgit -> fetchFromGitHub

zimbatm commented:
+100

offlinehacker commented:
nice :)

kamilchm commented:
I'm working on replacing all `fetchgit` with `fetchFromGitHub`. I don't know why but `fetchgit` doesn't work with version tags on travis :/
Give me an hour ;)

zimbatm commented:
Should go2nix be patched to automatically generate the right thing ?

kamilchm commented:
I left changes in `go2nix` until we settle on dependencies file format and `buildGoPackage.goDeps` attribute.

kamilchm committed:
Moving from fetchgit generated by go2nix to fetchFromGitHub where rev is
a tag

kamilchm commented on pkgs/top-level/all-packages.nix:
```diff
@@ -881,7 +881,7 @@ in
   mcrypt = callPackage ../tools/misc/mcrypt { };
 
-  mongodb-tools = goPackages.mongo-tools.bin // { outputs = [ "bin" ]; };
+  mongodb-tools = (callPackage ../tools/misc/mongodb-tools { }).bin;
```
Is it ok like this or maybe we want to produce every otutput? I couldn't figure it out and make 2 styles of it to make it consistent later.

kamilchm commented on pkgs/development/go-modules/generic/default.nix:
```diff
@@ -35,6 +40,31 @@ let
+  dep2src = goDep:
+    {
+      inherit (goDep) goPackagePath;
+      src = if goDep.fetch.type == "git" then
```
is there a better way to make some pattern matching on fetch type? Or maybe I need to change json file structure to fit better to nix?

cstrahan replied:
@kamilchm You can do something like:

```nix
let
  fetcher = {
    git = fetchgit;
    hg = fetchhg;
  }."${goDep.fetch.type}" or abort "Unrecognized package fetch type '${goDep.fetch.type}'";
```

kamilchm committed:
buildGo16Packages: fetchgit -> fetchFromGitHub

kamilchm commented:
strange thing happens on travis:
```
/run/user/1001/nix-build-consul-ui-0.6.4.drv-0/consul-v0.6.4-src/ui /run/user/1001/nix-build-consul-ui-0.6.4.drv-0/consul-v0.6.4-src
sh: /run/user/1001/nix-build-consul-ui-0.6.4.drv-0/consul-v0.6.4-src/ui/scripts/dist.sh: Permission denied
```

I can't reproduce it on my machine :/
What could be wrong?

groxxda commented:
Most certainly a travis issue you don't have to care about [See](https://github.com/NixOS/nixpkgs/pull/16022)

kamilchm commented:
So I leave travis as is.

kamilchm commented:
Trying again after @domenkozar fix

kamilchm commented:
Needs update after https://github.com/NixOS/nixpkgs/pull/16042

kamilchm committed:
buildGoPackage: use Go 1.6 by default

kamilchm committed:
nomad: extracted from goPackages

kamilchm committed:
goPackages: not needed now

kamilchm commented:
CC: @rushmorem

rushmorem commented:
@kamilchm awesome stuff you are doing here... Are you ready to have this merged?

kamilchm commented:
I'm not sure on ``deps.json`` structure. But if we agree that it's acceptable like it is, I can prepare updated ``go2nix`` to merge it as whole and refactor it later if needed.
One thing I'd like to have is a successful build on any CI. I don't know why travis is failing here :/ Can I use Hydra to make sure it wont brakes master after merging?

rushmorem commented on pkgs/applications/networking/cluster/nomad/default.nix:
```diff
@@ -0,0 +1,25 @@
+{ stdenv, lib, buildGoPackage, consul-ui, fetchFromGitHub }:
```
I don't see `lib` and `consul-ui` being used anywhere. Do you need these?

kamilchm replied:
ehh, quick copy from consul derivation :/ I'll fix it

zimbatm commented:
I tried running `nox-review pr 16017` but master is broken right now. I'm getting the following error:
`getting status of ‘/home/zimbatm/.nox/nixpkgs/pkgs/desktops/kde-5/frameworks-5.21’: No such file or directory`

zimbatm commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15890,7 +15878,7 @@ in
 
   kde5 =
     let
-      frameworks = import ../desktops/kde-5/frameworks-5.22 { inherit pkgs; };
```
looks like this slipped in during a rebase

kamilchm commented:
@zimbatm thx! Travis is green on my fork https://travis-ci.org/kamilchm/nixpkgs/builds/136399564 and we're back with strange `consul-ui` problem on travis:
```
sh: /run/user/1001/nix-build-consul-ui-0.6.4.drv-0/consul-v0.6.4-src/ui/scripts/dist.sh: Permission denied
```

zimbatm commented:
Running `nox-review` again I got a couple remaining errors:
```
error: build of ‘/nix/store/6px75ggl5kws4h8bw95sriv2qq56cnqc-php-imagick-3.1.2.drv’, 
‘/nix/store/8vrrn5iz9yisayl78fzjk3rfq2c97n5h-php-memcached-2.2.0.drv’, 
‘/nix/store/jpr1y2kxbszdfw3cszmgkngjnkrrgc3x-go1.6-git-lfs-1.2.0.drv’,
‘/nix/store/yva3qm0ff8x4crds7ixq6mdsmhcdkg7q-packer-0.10.1.drv’, 
‘/nix/store/z3g6dq00c0i69ky2yv3qg877h7nxa49h-go1.6-ipfs-i20160112--7070b4d.drv’ failed
```

rushmorem commented:
@kamilchm that permission error sounds to me like the bug that was in `nox-0.0.2` (https://github.com/madjar/nox/issues/30). It's suppossed to be fixed in `nox-0.0.3`. I'm not sure which version of `nox` the Travis scripts are pulling. If you can build fine locally with `nox-review` then you can safely ignore it.

kamilchm commented:
@zimbatm `git-lfs` fails on master too. I think its related to https://github.com/NixOS/nixpkgs/pull/16042. I reverted it to Go 1.5 for a quick fix.

zimbatm commented:
For packer I am seeing the following errors:
```
go/src/github.com/mitchellh/packer/packer-4e5f651/post-processor/amazon-import/post-processor.go:126: cannot use config (type *"github.com/mitchellh/packer/vendor/github.com/aws/aws-sdk-go/aws".Config) as type *"github.com/mitchellh/packer/packer-4e5f651/vendor/github.com/aws/aws-sdk-go/aws".Config in argument to "github.com/mitchellh/packer/packer-4e5f651/vendor/github.com/aws/aws-sdk-go/aws/session".New
```

rushmorem commented:
@kamilchm I don't think reverting `git-lfs` will fix this. Last time I checked `go-1.5` itself was broken. I don't know if it has been fixed yet. [As you can see on Hydra], `git-lfs` was broken on `go1.5` as well.

[As you can see on Hydra]: http://hydra.nixos.org/job/nixpkgs/trunk/git-lfs.x86_64-linux

kamilchm commented:
@zimbatm could try `git-lfs` and `packer` now?
@rushmorem is there a way to reproduce same `git-lfs` error on local machine or run hydra builds from my own fork?

rushmorem commented:
Nice thing about `nix-build` is that you can add `--keep-failed` or `-K` to keep the failed build directory in `/tmp` so you can look around to see what might have caused the failure.

kamilchm committed:
ipfs: build fails with Go 1.6, revert to 1.5

zimbatm committed:
goBuildPackage: export go

This is useful to make sure to use the same version of go in further
derivations.

zimbatm committed:
packer: fix compilation

* Remove duplicate "packer" source
* Use the same version of go everywhere

zimbatm commented:
@kamilchm fixed packer, look at the 3 last commits on my branch: https://github.com/zimbatm/nixpkgs/commits/pr/16017

kamilchm commented:
I found one more failing with 1.6 - `ipfs`.
@zimbatm packer alone needs its own `rework` story ;) I'll include your fixes here.

kamilchm commented:
@rushmorem I know `nix-build` and done it for all packages inluded here when I was extracting them. I was curious is there another easy to use tools than can save me from *works for me* syndrome :)

zimbatm committed:
git-lfs: 1.2.0 -> 2016-06-07

Fixes import issues after go1.5+

zimbatm commented:
@kamilchm yeah cherry-pick my changes. Just pushed a fix for git-lfs too.

rushmorem commented:
@kamilchm I see. In that case I can't think of anything else besides Hydra.

rushmorem commented:
`nox-review pr 16017` builds fine now. Thanks guys! Merging now...

kamilchm commented:
Thx! I'll create new PRs with updated `go2nix` and docs.

rushmorem commented:
@kamilchm That will be awesome. Thanks for all your work!

The following open pull requests will need to be revised and/or refactored after this https://github.com/NixOS/nixpkgs/pull/15914 https://github.com/NixOS/nixpkgs/pull/13819 https://github.com/NixOS/nixpkgs/pull/13566 https://github.com/NixOS/nixpkgs/pull/12010. Adding them here to bring this pull request to their authors' attention...

rushmorem commented:
@kamilchm I just want to point out that the reason why Travis may pass on your branch but fail on your pull request is because it does further checks for pull requests. At first Travis will only build packages that your changes touch on but if it's a pull request then it goes on to build all the packages that depend on the packages you changed (using `nox-review`).

rushmorem commented:
[Documentation](http://hydra.nixos.org/build/36798894/download/1/nixpkgs/manual.html#sec-language-go)...

lethalman commented:
Not very happy with this change. It introduces a new way to specify dependencies with a different format other than nix, and duplicates a lot of similar go packages in the system.

But I see it's been agreed by quite some people, hopefully it goes well.

kamilchm commented:
@lethalman do you see anything that can we can do to make `goDeps` better? We can change the format but I thought that having `fromJSON` in Nix makes it almost first class citizen? It also easier to automate things when you can use JSON. As I said I wasn't sure with the exact structure of `goDeps` and I can refactor it if we can find better format.
We can reuse duplicated dependencies with `goDeps.include` like I did initialy. I think it's more about the discussion if we want to be sure we build the exact same package that was approved by upstream (https://github.com/NixOS/nixpkgs/pull/16172#issuecomment-225437289).
On the other hand it looks like `goPackages` got leaner overall (PR diff +3,822 −4,509) and give us *better structure*, more space for package metadata and it looks like it should be a lot easier to found package maintainers now.
I'm really open for any improvements as it all was made for better packaging in mind.

zimbatm commented:
I'm not a big fan of the common libs.json.

nix already de-duplicates the derivations if any similar dependency is shared (and it's only at build time). And git already compresses the code in the pack files. I think we'll find out that libs.json is more maintenance that it's really worth but we'll see.

lethalman commented:
@kamilchm I don't like the idea that sharing libraries is killed, and that reusing libraries potentially need special metadata in json.

How does it work with e.g. a library that needs a C dependency?

Where's it written that it's a better structure and that it gives more space to metadata?

lethalman commented:
@kamilchm I don't understand why the `toJSON/fromJSON` instead of plain .nix? Nix files can be easily generated from tools as well, except that with nix you have native support for imports instead of reimplementing it in nix itself.

kamilchm commented:
The `libs.json` was simple succession from one big `go-packages.nix`.
JSON was influenced by https://github.com/NixOS/nixpkgs/pull/13819. We can change it to Nix but that could get us back to `go-packages.nix` and if we want to reuse dependencies in one big file where people do many manual changes this can't be easily automated with Go which is the best for automation in Go ecosystem. Could you provide an example of refactoring one simple Go package done with only Nix that could drive us to better direction?
Do you have an example when you can't share a library now? For a C dependency you need to add it at `buildGoPackage` level like https://github.com/kamilchm/nixpkgs/blob/a5a1d45636e87e8872d94d6e86e43af44eb8c3ba/pkgs/tools/X11/go-sct/default.nix#L18
It may not be the best but it doesn't look like a big duplication with current set of packages. Nevertheless, as I said I'm open for ideas. 
Saying about structure and metadata I mean that it should encourage people to provide better looking packages when you're doing `pks/servers/monitoring/something/default.nix` and it will be your own package to maintain than throwing few new lines surrounded with almost no metadata in `go-packages.nix`.

lethalman commented:
@kamilchm a .nix is quite similar to a .json, it shouldn't be an issue to generate it with go. A .nix is better in terms of operations, like calling nix functions, whereas with json you'd actually need to interpret them with nix itself.

edolstra commented:
:+1: on getting rid of `go-packages.nix`, but :-1: on the JSON stuff. `goDeps` could easily be expressed as a attribute set, obviating the need for features such as `include`, which are already available in the Nix language. (Actually it occurs to me that `pkgs/development/go-modules/libs.json` is really `go-packages.nix` in disguise.)

kamilchm commented:
If someone could show one simple Go package with its dependencies in Nix as an example I can refactor all `libs.json` automatically to the form that will be acceptable.
I'll think about it too and try to prepare another proposal later this week.

chris-martin commented:
> I started with an assumption that there's no need to have `goPackages` that aren't used in a compiled program.

IMO this is a pretty bad assumption. Removing `goPackages` has broken my local packages that used Go libraries, and it removed the `gore` and `go-ethereum` packages which produced executables.

chris-martin commented:
I'm concerned that writing the dependency logic in JSON instead of Nix makes it difficult for multiple revisions of a dependency to coexist. Is there an example of that anywhere?

kamilchm commented:
@chris-martin I move general discussion to the mailing list.
I can't find usage of `gore` and `go-ethereum` packages anywhere before this PR. Is it related to your local packages? I could restore it as `top-level` packages.

chris-martin commented:
@kamilchm Yeah, those should be restored as top-level packages.

chris-martin commented:
Mailing list link: https://www.mail-archive.com/nix-dev@lists.science.uu.nl/msg20641.html

chris-martin commented:
I've submitted #16353 to restore `go-ethereum`.

dguibert commented:
By getting rid of go-packages.nix, rclone has removed (introduced by #15852).

I wanted to use `go2nix save` to rebuild the derivation. But how can I populate GOPATH with all the dependencies for rclone (on NixOS)?

kamilchm commented:
It should be as easy as using `go get github.com/ncw/rclone` as stated in http://rclone.org/install/, then go to `$GOPATH/src/github.com/ncw/rclone` and execute `go2nix save` and you should see 2 files there `default.nix` and `deps.json`. If something goes wrong it'll be `go2nix` bug that you can report there.

kamilchm commented:
@edolstra is there a way to use nix parser as a library? I can generate nix instead of json using text templates in Go but I can't automate package reuse in go2nix without parsing nix.
I remember that there was an initiative by @zimbatm  to make a subset of nix with only literal values, but it looks like it's ended with JSON anyway https://github.com/ryantrinkle/nixpkgs/blob/be30ba8e0eabbc2483122ddf9d5911484df3a578/pkgs/build-support/fetchgit/nix-prefetch-git#L326 ?

zimbatm commented:
@kamilchm I think we should drop the libs.json altogether. Then it's not a problem because you don't have to parse the nix code, you just generate new deps.nix files.

Because of the content-addressable store, if two programs have the same dependency it will already result in the same derivation in the store. Git also has compression in the pack files so it won't make much difference to duplicate the dependencies on disk. And finally most users will just use the binary builds so it won't make any differences to them.

kamilchm commented:
You're right. I'll make a proposal PR with `deps.nix` and leave the package reuse at nix store level.

copumpkin commented:
@kamilchm did you lose all the `fetchFromGitHub` calls in this change? And the `buildFromGitHub` call was quite nice, since the vast majority of Go packages have a ton of nearly identical boilerplate and fetch from github.

copumpkin commented:
@kamilchm @zimbatm I don't think sharing is a storage-level concern for most of us, so the content-addressable deduplication doesn't really address my concern with it. It's a human-level concern for me, where I want to make sure that if I find that I needed to patch some particular Go package for use with Nix (perhaps because it assumed silly things about an FHS filesystem), I don't have to do that every time I use it, but can do that once. Or that if a particular Go package ends up having a security vulnerability, that I don't need to go hunt down a thousand libs.json files to figure out who's vulnerable.

So in some sense I'm kind of disappointed that this is now the approach, and wish I'd paid attention while this was happening.
